ferrocene/tools/self-test/src/linkers.rs

```diff
@@ -1,71 +1,257 @@
+                        ));
+                        let mut need_no_lto = false;
+                        for arg in args {
+                            if std::env::var("FST_PRINT_LINKER_ARGS").is_ok() {
+                                reporter
+                                    .note(&format!("`{compiler_name}` provides argument {arg:?}"));
+                            }
+                            if arg.contains("liblto_plugin") {
+                                need_no_lto = true;
+                            }
+                        }
+                        target.rustflags.push(format!("-Clinker={compiler_name}"));
+                        if need_no_lto {
+                            target.rustflags.push(format!("-Clink-arg=-fno-lto"));
+                        }
+                        continue 'target_loop;
+                    }
+                    Err(e) => {
+                        // Try again until we run out of compilers
+                        if std::env::var("FST_PRINT_DETAILED_ERRORS").is_ok() {
+                            reporter.note(&format!("`{compiler_name}` failed with {e}"));
+                        }
+                    }
+                }
             }
         }
+        return Err(Error::SuitableCCompilerNotFound { target: target.triple.to_owned() });
     }
 
     Ok(())
 }
 
-fn check_bundled_lld(
-    reporter: &dyn Reporter,
-    sysroot: &Path,
-    targets: &[String],
-) -> Result<PathBuf, Error> {
+/// Check if the given system C compiler works.
+///
+/// We are given a path to the real `ld.lld`, but we also test it with a fake
+/// `ld.lld` as well.
+///
+/// We are also given a path to a fresh temporary directory we can put source
+/// code in.
+///
+/// Returns the path to the C compiler, and a list of arguments that the C
+/// compiler gives to the linker.
+fn check_system_compiler(
+    environment: &Environment,
+    target: &str,
+    compiler_name: &str,
+    lld_dir: &Path,
+    temp_dir: &Path,
+) -> Result<(PathBuf, Vec<String>), Error> {
+    let cc_path = find_binary_in_path(environment, &compiler_name)
+        .map_err(|error| Error::CCompilerNotFound { name: compiler_name.to_owned(), error })?;
+
+    // Part 1. Check with the real ld.lld - can we make a binary?
+
+    compile_c_program(&cc_path, lld_dir, temp_dir)?;
+
+    // Part 2. Make a fake linker, and get GCC to try and use it. What arguments
+    // does it give our fake linker?
+
+    let args_file = make_fake_linker(temp_dir)?;
+
+    let linker_args = check_compiler_linker_args(target, &cc_path, temp_dir, &args_file)?;
+
+    Ok((cc_path, linker_args))
+}
+
+/// Check if the given system C compiler works.
+///
+/// We are given a path to the real `ld.lld`.
+///
+/// We are also given a path to a fresh temporary directory we can put source
+/// code in.
+///
+/// Returns the path to the C compiler.
+fn compile_c_program(cc_path: &Path, lld_dir: &Path, temp_dir: &Path) -> Result<(), Error> {
```

Comment:
I think this is redundant with `make_fake_linker`? That also compiles a C program, and we can pass `-fuse-ld`/`-B` to it as well.

Reply:
I actually thought the same thing, but then I realised `make_fake_linker` needs to make a host binary, but `compile_c_program` needs to do a cross-compilation. I should clarify this in the comments.

Reply:
You're right, good catch.